src/unix/blockq.c

```diff
@@ -294,80 +213,80 @@ void kern_blockq_handle_completion(blockq bq, u64 bq_flags, io_completion comple
 }
 KLIB_EXPORT_RENAME(kern_blockq_handle_completion, blockq_handle_completion);
 
-/* XXX This deserves another pass; blockq_item should really just be embedded into thread. */
-boolean blockq_flush_thread(blockq bq, thread t)
-{
-    boolean unblocked = false;
-    blockq_debug("bq %p, name %p\n", bq, blockq_name(bq));
-
-    /* XXX take irqsafe spinlock */
-
-    list_foreach(&bq->waiters_head, l) {
-        blockq_item bi = struct_from_list(l, blockq_item, l);
-        if (bi->t != t)
-            continue;
-
-        blockq_apply_bi_locked(bq, bi, BLOCKQ_ACTION_BLOCKED | BLOCKQ_ACTION_NULLIFY);
-        unblocked = true;
-    }
-
-    /* XXX release lock */
-    return unblocked;
-}
-
 /* Wake all waiters and empty queue, typically for error conditions,
    closed pipe/connections, etc. Actions are called with nullify set,
    indicating the last time that the action will be used by the
    blockq, regardless of what the action returns.
 */
-
 void blockq_flush(blockq bq)
 {
     blockq_debug("bq %p - \"%s\"\n", bq, blockq_name(bq));
-
-    /* XXX take irqsafe spinlock */
-    list_foreach(&bq->waiters_head, l) {
-        blockq_item bi = struct_from_list(l, blockq_item, l);
-        blockq_apply_bi_locked(bq, bi, BLOCKQ_ACTION_BLOCKED | BLOCKQ_ACTION_NULLIFY);
-    }
-    /* XXX release lock */
+    do {
+        u64 saved_flags = spin_lock_irq(&bq->lock);
+        list l = list_get_next(&bq->waiters_head);
+        if (!l) {
+            spin_unlock_irq(&bq->lock, saved_flags);
+            return;
+        }
+        list_delete(l);
+        spin_unlock_irq(&bq->lock, saved_flags);
+        blockq_flush_thread(bq, struct_from_list(l, thread, bq_l));
```

Comment:
blockq_flush_thread() calls blockq_wake_thread_internal(), which asserts that `(t->bq_l.next && t->bq_l.prev)` (and also tries to remove the thread from the waiter list). I think we should just call blockq_apply() instead of blockq_flush_thread().

Reply:
Ok, I changed the call to blockq_flush_thread to a call to blockq_apply. Thanks.